HISTORY.md

```diff
@@ -12,6 +13,7 @@
 ### Public API changes
 * Add rollback_deletion_type_callback to TransactionDBOptions so that write-prepared transactions know whether to issue a Delete or SingleDelete to cancel a previous key written during prior prepare phase. The PR aims to prevent mixing SingleDeletes and Deletes for the same key that can lead to undefined behaviors for write-prepared transactions.
 * EXPERIMENTAL: Add new API AbortIO in file_system to abort the read requests submitted asynchronously.
+* CompactionFilter::Decision has a new value: kRemoveWithSingleDelete. If CompactionFilter returns this decision, then CompactionIterator will use `SingleDelete` to mark a key as removed.
```

Comment:
Sorry I don't understand. Does this only work when there are two PUTs associated with a key? The newer PUT is converted to SINGLEDEL, which then covers the older PUT. Otherwise what would the newer PUT cover?

The last time we considered this the plan was to have a decision like `kPurge` that purges the key without adding any tombstone for cases where only one version of the key is known to exist.

Reply:
> Does this only work when there are two PUTs associated with a key?

If there can be multiple PUTs for the same key, compaction filter should never return `kRemoveWithSingleDelete`.

Reply:
If application does not intend to use SingleDelete for certain keys, then the compaction filter should always use `kRemove` to remove data.
If application does use SingleDelete for certain keys, then these keys should always be removed by compaction filter via `kRemoveWithSingleDelete`. In this case, there should not be two consecutive Puts.

Reply:
> > Does this only work when there are two PUTs associated with a key?
> 
> If there can be multiple PUTs for the same key, compaction filter should never return `kRemoveWithSingleDelete`.

That's good and intuitive. Then when should it ever return `kRemoveWithSingleDelete`? If there's one PUT for the key, converting it to a SINGLEDEL will cover nothing, so I don't see a need to write it.

Reply:
Compaction filter (provided by user) should determine when to return `kRemoveWithSingleDelete`. If it is returned, do you mean we can just drop the Put instead of writing a SingleDelete? I think so.

Reply:
> If application does use SingleDelete for certain keys, then these keys should always be removed by compaction filter via `kRemoveWithSingleDelete`.

How about  compaction filter does not remove them at all? I don't think `SingleDelete()` guarantees it works when SINGLEDELs can cover nothing, right?

Reply:
> If it is returned, do you mean we can just drop the Put instead of writing a SingleDelete?

Yes. I also think it is a bug to write a SINGLEDEL covering nothing, which makes this feature impossible to use correctly

Reply:
Do you mean a single `SingleDelete` is a bug? Also what about a number of consecutive single deletes?
I thought compaction iterator can handle this case. Current write-prepared transaction can also lead to this due to snapshot release.

Reply:
> Do you mean a single `SingleDelete` is a bug? Also what about a number of consecutive single deletes? I thought compaction iterator can handle this case.  Current write-prepared transaction can also lead to this due to snapshot release.
> UPDATE:
> 
> > I don't think SingleDelete() guarantees it works when SINGLEDELs can cover nothing, right?
> 
> Can you elaborate on this?

I did not know `CompactionIterator` can handle this. I thought this (https://github.com/facebook/rocksdb/blob/06394ff4e77a4d1d1ad34539148a5ecd02fc9791/include/rocksdb/db.h#L433-L434) means there must be exactly one PUT, not up to one PUT. We can check how CompactionIterator handles two consecutive SINGLEDELs.

Reply:
I think multiple SD will be skipped and the earliest one will be considered:
https://github.com/facebook/rocksdb/blob/main/db/compaction/compaction_iterator.cc#L636

Reply:
Got it, added #9930 to test it.

Back to the original question then - is there a use case it's desirable to convert to SINGLEDEL covering nothing rather than purging?

Reply:
Not considering write conflict detection snapshot, I guess no, but this is already not well-supported with compaction filters. I will revisit this once all stress tests become stable.
Even so, I still think the compaction filter needs to return a different `Decision` then `kRemove` so that compaction iterator can determine the PUT can just be dropped instantly.